crates/loans/src/lib.rs

```diff
@@ -0,0 +1,2055 @@
+// Copyright 2022 Interlay.
+// This file is part of Interlay.
+
+// Copyright 2021 Parallel Finance Developer.
+// This file is part of Parallel Finance.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Loans pallet
+//!
+//! ## Overview
+//!
+//! Loans pallet implement the lending protocol by using a pool-based strategy
+//! that aggregates each user's supplied assets. The interest rate is dynamically
+//! determined by the supply and demand.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::rate_model::*;
+
+use currency::Amount;
+use frame_support::{
+    log,
+    pallet_prelude::*,
+    require_transactional,
+    traits::{
+        tokens::fungibles::{Inspect, Mutate, Transfer},
+        UnixTime,
+    },
+    transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use num_traits::cast::ToPrimitive;
+use orml_traits::MultiCurrency;
+pub use pallet::*;
+use pallet_traits::{
+    ConvertToBigUint, LoansApi as LoansTrait, LoansMarketDataProvider, MarketInfo, MarketStatus, PriceFeeder,
+};
+use primitives::{Balance, CurrencyId, Liquidity, Price, Rate, Ratio, Shortfall, Timestamp};
+use sp_runtime::{
+    traits::{
+        AccountIdConversion, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, SaturatedConversion, Saturating,
+        StaticLookup, Zero,
+    },
+    ArithmeticError, FixedPointNumber, FixedU128,
+};
+use sp_std::{marker, result::Result};
+
+pub use orml_traits::currency::{OnDeposit, OnSlash, OnTransfer};
+use sp_io::hashing::blake2_256;
+pub use types::{BorrowSnapshot, EarnedSnapshot, Market, MarketState, RewardMarketState};
+pub use weights::WeightInfo;
+
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod farming;
+mod interest;
+#[cfg(test)]
+mod lend_token;
+mod rate_model;
+mod types;
+
+pub mod weights;
+
+type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+type AssetIdOf<T> = <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+type BalanceOf<T> = <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub struct OnSlashHook<T>(marker::PhantomData<T>);
+impl<T: Config> OnSlash<T::AccountId, AssetIdOf<T>, BalanceOf<T>> for OnSlashHook<T> {
```

Comment:
Add TODO note for link to orml-tokens update, also for fallible change on substrate side